Convert an assertion to if-condition in `ReplaceSelectionCommand::doApply()`

Ahmad-S792 opened the pull request:
#### cb8d258708a5e61e22b9d3f6de9fd518041d5b91
<pre>
Convert an assertion to if-condition in `ReplaceSelectionCommand::doApply()`

<a href="https://bugs.webkit.org/show_bug.cgi?id=127684">https://bugs.webkit.org/show_bug.cgi?id=127684</a>
<a href="https://rdar.apple.com/problem/27685262">rdar://problem/27685262</a>

Reviewed by Ryosuke Niwa.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/8f313902e23e2589ddd9d3d2bfed8ee61e979f17">https://chromium.googlesource.com/chromium/src.git/+/8f313902e23e2589ddd9d3d2bfed8ee61e979f17</a>

This patch changes following assertion:

`ASSERT(insertionBlock != currentRoot)`

to if-condition in `ReplaceSelectionCommand::doApply()`, since we found a
pattern to hit this assertion in &quot;paste-text-009.html&quot;.

This test changes current test to use &apos;testharness&apos; script but instead of full
merge from Blink commit, this limits it to only test change (i.e., 009.html).
Only change is &apos;root&apos; editable:

&quot;past-text-009.html&quot;: (testharness)
    |currentRoot| is &quot;&lt;div&gt;one&quot;
    | insertionBlock | is &quot;&lt;div&gt;one&quot;

  &quot;paste-text-009.html&quot;:
    |currentRoot| is BODY
    | insertionBlock | is &quot;&lt;div&gt;Omitted&quot;.

* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(ReplaceSelectionCommand::doApply):
* LayoutTests/editing/pasteboard/paste-text-009.html: Rebaselined
* LayoutTests/editing/pasteboard/paste-text-009-expected.txt: Add Test Expectation
* LayoutTests/platform/gtk/editing/pasteboard/paste-text-009-expected.png: Removed
* LayoutTests/platform/gtk/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.png: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/mac/editing/pasteboard/paste-text-009-expected.png: Ditto
* LayoutTests/platform/mac/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/wincairo/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.txt: Add Platform Specific Expectation
* LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash.html: Add Test Case
* LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/277291@main">https://commits.webkit.org/277291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/356d519bb0466c8ae93638865c498159c2709c5a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26368 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49872 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43238 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49496 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23828 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38440 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19752 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41824 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5233 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43578 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42232 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51748 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22216 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18589 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45735 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23491 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44743 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10414 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dd0e9c1b914b211375e25688cd983e91ee0f8695. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26896

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b33d7a121ca5265acaeaec72cc4ba263b1c4ce5e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26896

Ahmad-S792 commented:
@rniwa - I added 'crash' (assert) test as well now. :-)

rniwa commented on LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash-expected.txt:
```diff
@@ -0,0 +1,8 @@
+layer at (0,0) size 800x600
```
Please call testRunner.dumpAsText so that the output will be plain text.

Ahmad-S792 replied:
Done. :-) Thanks for your guidance, learnt something new. :-)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/356d519bb0466c8ae93638865c498159c2709c5a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26896

Ahmad-S792 commented on LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash.html:
```diff
@@ -0,0 +1,17 @@
+<body onload="
+	if (document.counter)
```
Done. :-)

Commit:
Convert an assertion to if-condition in `ReplaceSelectionCommand::doApply()`

https://bugs.webkit.org/show_bug.cgi?id=127684
rdar://problem/27685262

Reviewed by Ryosuke Niwa.

Merge: https://chromium.googlesource.com/chromium/src.git/+/8f313902e23e2589ddd9d3d2bfed8ee61e979f17

This patch changes following assertion:

`ASSERT(insertionBlock != currentRoot)`

to if-condition in `ReplaceSelectionCommand::doApply()`, since we found a
pattern to hit this assertion in "paste-text-009.html".

This test changes current test to use 'testharness' script but instead of full
merge from Blink commit, this limits it to only test change (i.e., 009.html).
Only change is 'root' editable:

"past-text-009.html": (testharness)
    |currentRoot| is "<div>one"
    | insertionBlock | is "<div>one"

  "paste-text-009.html":
    |currentRoot| is BODY
    | insertionBlock | is "<div>Omitted".

* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
(ReplaceSelectionCommand::doApply):
* LayoutTests/editing/pasteboard/paste-text-009.html: Rebaselined
* LayoutTests/editing/pasteboard/paste-text-009-expected.txt: Add Test Expectation
* LayoutTests/platform/gtk/editing/pasteboard/paste-text-009-expected.png: Removed
* LayoutTests/platform/gtk/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.png: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/mac/editing/pasteboard/paste-text-009-expected.png: Ditto
* LayoutTests/platform/mac/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/wincairo/editing/pasteboard/paste-text-009-expected.txt: Ditto
* LayoutTests/platform/ios/editing/pasteboard/paste-text-009-expected.txt: Add Platform Specific Expectation
* LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash.html: Add Test Case
* LayoutTests/editing/execCommand/canvas-counter-rooteditable-crash-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/277291@main

webkit-commit-queue commented:
Committed 277291@main (cb8d258708a5): <https://commits.webkit.org/277291@main>

Reviewed commits have been landed. Closing PR #26896 and removing active labels.
